### ISSUE TITLE
set mode desperate and show mode desperate problems with large values

### ISSUE THREAD

**Issue by karlmsmith** [NOAA-PMEL/Ferret#1091]
**Reported by karl smith on 23 Mar 2011 17:38 UTC**
``` set mode desperate `96 * 1024 * 1024` ``` fails, but ``` set mode desperate 1234567890123456789 ``` apparently succeeds.  But the value assigned is unknown because the display for large values is
```
yes?  show mode desp
      MODE            STATE        ARGUMENT
      DESPERATE        SET         *******
```

Migrated-From: http://dunkel.pmel.noaa.gov/trac/ferret/ticket/1819

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 23 Mar 2011 19:01 UTC**
This looks like one of the instances we've been running into where old assumptions about the sizes of things are breaking down with what's available on modern computers.

The argument to MODE DESPERATE is read from the command-line buffer as an integer using an I10 format which means that 1234567890123456789 is read as 1234567890. Even if we did read the number correctly, we'd run into the same kind of integer overflow trouble we did with SET MEMORY and which Karl has fixed.

In SHOW MODE, The value of MODEs is listed using an I7 format, for a nice formatted listing. That can be adjusted.

Now, for the case of "set mode desperate `96*1024*1024`", the command-parsing step has replaced `96*1024*1024` with the floating point value, 1.006633E+08  and then that can't be handled by the code that tries to read the argument for MODE DESPERATE as an integer.  Here, if there's an error reading as an integer I could try to read it as a floating-point value and convert it an integer, so long as the value isn't too big.

**Comment by karlmsmith** [comment]
**Comment by steven.c.hankin on 23 Mar 2011 19:40 UTC**
Would this work?  (The thing you suggest in your final line.)
  1.  read the value into a float using a "*" format
  2.  check that its value is plausible -- error if not
  3. convert to integer

Then you can be sure you'll never get an argument value too big to print out with the desired format.

Unanswered above:  how does it happen that the default value of MODE DESP prints as "***"?

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 23 Mar 2011 19:51 UTC**
Those steps sound just right.


The default value of MODE DESPERATE is printed correctly.  From a fresh Ferret session,

```
yes? set mode desperate
yes? sho mode desperate
      MODE            STATE        ARGUMENT
      DESPERATE        SET         2560000
```

But once a setting has been made, its value is kept:

```
yes? set mode desperate 4000000
yes? sh mode desperate
      MODE            STATE        ARGUMENT
      DESPERATE        SET         4000000

yes? cancel mode desperate
yes? sho mode desperate
      MODE            STATE        ARGUMENT
      DESPERATE     CANCELLED      4000000
```


and so, once you had made a setting that was out of range of the format statement, I suspect that you continued to "see stars"?  

Should CANCEL MODE DESPERATE revert the value of the mode setting back to the default?  I think that might be more reasonable.

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 23 Mar 2011 20:02 UTC**
I said [CANCEL MODE DESPERATE revert the value of the mode setting back to the default? I think that might be more reasonable.](Should)

Karl pointed out that the value of MODE DESPERATE changes with SET MEMORY. "Whenever the size of memory is set using SET MEMORY the MODE DESPERATE argument is reset at one tenth of memory size." So we could revert to that value on a CANCEL MODE DESPERATE, or just leave it at the previous valid setting, now that we're not going to make invalid settings.

**Comment by karlmsmith** [comment]
**Comment by steven.c.hankin on 23 Mar 2011 20:40 UTC**
Karl's point explains why the default value of MODE DESP displayed stars for me earlier;  I had bumped the value of MEMORY to 2000.

My inclination would be to leave the resetting behavior of the MODE DESPERATE argument value as-is.  Better not to risk backwards compatibility on someone's script, unless we see a compelling reason to do so.

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 25 Mar 2011 17:30 UTC**
Fixed as outlined in comment #2 above, without making any changes to the behavior on CANCEL MODE DESPERATE.  See gnl/get_mode_arg.F  xeq/xeq_show.F
